lib/grpnames.gi

```diff
@@ -709,56 +710,137 @@ InstallGlobalFunction(DirectFactorsOfGroupKN,
+    G := arg[1];
+    method := arg[2];
+    fullNs := true;
+  elif Length(arg) = 2 and IsGroup(arg[1]) and IsList(arg[2])
+      and ForAll( Set(arg[2]), N ->
+                              IsSubgroup(arg[1], N) and IsNormal(arg[1], N))
+      then
+    G := arg[1];
+    Ns := ShallowCopy(arg[2]);
+    fullNs := false;
+  elif Length(arg) = 3 and IsGroup(arg[1]) and IsList(arg[2])
+      and ForAll( Set(arg[2]), N ->
+                              IsSubgroup(arg[1], N) and IsNormal(arg[1], N))
+      and arg[3] in ["all", "any", "str"] then
+    G := arg[1];
+    Ns := ShallowCopy(arg[2]);
+    method := arg[3];
+    fullNs := false;
+  else
+    Error("usage: SemidirectDecompositionsOfFiniteGroup(<G> [, <Ns>] [, <mthd>])");
+  fi;
 
-    # non-1-or-G normal subgroups
-    Ns := Reversed( Filtered(Hs, H -> IsNormal(G, H)) );
+  if HasSemidirectDecompositions(G) then
+    NHs := [ ];
+    for NH in SemidirectDecompositions(G) do
+      N := NH[1];
+      H := NH[2];
+      if method in [ "any", "str" ] and not IsTrivial(N)
+        and not IsTrivial(H) and
+        ( not IsBound(Ns) or N in Ns ) then
+        return [ N, H ];
+      elif method="all" and
+        ( not IsBound(Ns) or N in Ns ) then
+        AddSet(NHs, [ N, H ]);
+      fi;
+    od;
+    if method in [ "any", "str" ] then
+      return fail;
+    elif method = "all" then
+      return NHs;
+    fi;
+  fi;
 
-    # find first decomposition
-    firstHN := function ()
+  if method in [ "any", "str" ] then
+    if HasSemidirectProductInfo(G) then
+      N := Image(Embedding(G, 2));
+      H := Image(Embedding(G, 1));
+      if not IsTrivial(N) and not IsTrivial(H) and
+        ( not IsBound(Ns) or N in Ns ) then
+        return [ N, H ];
+      fi;
+    fi;
+    N := NormalHallSubgroupsFromSylows(G, "any");
+    if N <> fail then
+      # by the Schur-Zassenhaus theorem there must exist a complement
+      if method = "any" then
+        H := ComplementClassesRepresentatives(G, N)[1];
+        return [ N, H ];
+      # only the isomorphism type of the complement is interesting
+      elif method = "str" then
+        Assert(1, Length( ComplementClassesRepresentatives(G, N) ) > 0);
+        return [ N, G/N ];
+      fi;
+    fi;
+  fi;
 
-      local H, N, sizeNs;
+  # simple groups have no nontrivial normal subgroups
+  if IsSimpleGroup(G) then
+    if method in [ "any", "str" ] then
+      return fail;
+    elif method = "all" then
+      return [ [TrivialSubgroup(G), G], [G, TrivialSubgroup(G)] ];
+    fi;
+  fi;
 
-      sizeNs := List(Ns, Size);
-      for H in Hs do
-        if Size(G)/Size(H) in sizeNs then
-          for N in Filtered(Ns, N -> Size(N) = Size(G)/Size(H)) do
-            if IsTrivial(NormalIntersection(N, H)) then
-              return Size(H);
-            fi;
-          od;
-        fi;
-      od;
-      return 0;
-    end;
+  if not IsBound(Ns) then
+    Ns := ShallowCopy(NormalSubgroups(G));
```

Comment:
@hungaborhorvath Thanks for the rebase.

As long as packages add new methods for things like `NormalSubgroups`, we will never be able to guarantee that these descriptions remain stable. And of course you are right, fully ordering the normal subgroups would be foolish in general.

So, perhaps we should simply stop any pretense that there is stability, and thus remove the relevant tests. Or only run them when the right packages are (not) loaded.

Another option would be to add another option, say `stable` or `fullyinvariant` or so to `StructureDescription` which instructs it to sort `NormalSubgroups` (assuming that is even enough). Then, compute the precomputed descriptions using that option, and use that option for the tests that verify the precomputed values. And indeed, do not run this test as part of the standard test suite. (I am not even sure it should be a regular test... Perhaps it would be better to have a function which (re)generates the file containing the precomputed names. Then, one can run that manually from time to time, if one is concerned about behavior changes).

Tests which fail at the slightest change of things are useless for us; esp. if mathemtically valid improvements cause these failures. Also, we never guarantee that structure descriptions are unique or never change, do we?